src/ref_count.rs

```diff
@@ -0,0 +1,370 @@
+                    // happens in scenario 1 during successful remove_if_last.
+                    // happens in scenario 1 while dropping the interner.
+                    // does not happen during interning race because there ref count is 1
+                    #[cfg(feature = "println")]
+                    println!("null {:p} {:p}", self, *self);
+                    break;
+                } else {
+                    // we got a valid pointer because our weak reference is still in place
+                    //
+                    // THIS FUNCTION POINTER MUST BE USED SUCCESSFULLY EXACTLY ONCE!
+                    //
+                    let raw_arc = remove_ptr as *const ();
+                    let remove_if_last = unsafe { *remove_ptr };
+                    match remove_if_last(raw_arc, self) {
+                        RemovalResult::Removed => {
+                            // scenario 1 confirmed and races won:
+                            // weak reference on interner has been dropped, so has our ref_count
+                            // so the code below will now drop this value
+                            #[cfg(feature = "println")]
+                            println!("removed {:p} {:p}", self, *self);
+                            break;
+                        }
+                        RemovalResult::NotRemoved => {
+                            // scenario 1 confirmed and race lost:
+                            // another thread has won the race and obtained a fresh reference, so we
+                            // keep our weak reference and put back the removal function pointer
+                            self.inner().remove_if_last.store(remove_ptr, Release);
+                            // at this point it is unclear what else has happened (other reference could already
+                            // have been dropped or interner dropped), so we must check again: seeing 3 means that
+                            // someone else will successfully use the pointer in the future, seeing 1 means that
+                            // the situation has been cleared up permanently, seeing 2 we need to try again
+                            #[cfg(feature = "println")]
+                            println!("loop {:p} {:p}", self, *self);
+                        }
+                        RemovalResult::MapGone => {
+                            // scenario 2 or scenario 1 with concurrent drop of interner
+                            // the interner has begun dropping at some point in the past, so its reference to
+                            // us is either gone or will be gone soon; in any case our weak reference to it is toast
+                            #[cfg(feature = "println")]
+                            println!("gone{:p} {:p}", self, *self);
+                            break;
+                        }
+                    }
+                }
+            } else {
+                break;
+            }
+        }
```

Comment:
How about putting this first?
```rs
loop {
  if self.inner().refs.load(Relaxed) != 2 {
    break;
  }
  …
}
```
Saves the majority of the code 1 level of indentation.